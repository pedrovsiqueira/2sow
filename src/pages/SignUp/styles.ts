import styled from 'styled-components';

export const Container = styled.div`
  width: 100vw;
  min-height: 100vh;
  margin-top: 82px;

  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  figure {
    text-align: center;
  }

  img {
    background: #fefefe;
    box-shadow: 18px 18px 30px #d1d9e6, -18px -18px 30px #ffffff;
    border-radius: 14px;
    padding: 21px 58px;
  }

  section {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: flex-start;

    label {
      font-size: 2rem;
      font-weight: 600;
      letter-spacing: 0.02em;
      font-variant: small-caps;
      margin: 24px 0 8px 0;
    }
  }

  div {
      display: flex;
      justify-content: center;
  }
`;
