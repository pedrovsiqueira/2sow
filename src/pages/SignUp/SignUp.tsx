import React, { useCallback, useRef } from 'react';
import Input from '../../components/Input/Input2';
import Button from '../../components/Button/Button';
import { Container, GoBackStyled, ButtonSection } from './styles';
import { FormHandles } from '@unform/core';
import { Form } from '@unform/web';
import * as Yup from 'yup';
import getValidationErrors from '../../utils/getValidationErrors';
import logo from '../../assets/logo.svg';
import { Link, useHistory } from 'react-router-dom';
import api from '../../services/api';
import { useToast } from '../../hooks/toast';
import backIcon from '../../assets/back-icon.svg';

interface SignUpFormData {
  email: string;
  confirmEmail: string;
  password: string;
}

const SignUp: React.FC = () => {
  const formRef = useRef<FormHandles>(null);
  const { addToast } = useToast();
  const history = useHistory();

  const handleSubmit = useCallback(
    async (data: SignUpFormData) => {
      try {
        formRef.current?.setErrors({});

        const schema = Yup.object().shape({
          email: Yup.string()
            .required('Confirme seu e-mail')
            .email('Digite um e-mail válido'),
          confirmEmail: Yup.string()
            .required('Confirme seu e-mail')
            .email('Digite um e-mail válido')
            .oneOf([Yup.ref('email')], 'Emails precisam ser iguais'),
          password: Yup.string().min(5, 'No mínimo 5 digitos'),
        });

        await schema.validate(data, {
          abortEarly: false,
        });

        await api.post('/users', data);

        history.push('/login');

        addToast({
          type: 'success',
          title: 'Cadastro realizado!',
          description: 'Você já pode fazer o login',
        });
      } catch (err) {
        if (err instanceof Yup.ValidationError) {
          const errors = getValidationErrors(err);
          formRef.current?.setErrors(errors);

          return;
        }

        addToast({
          type: 'error',
          description: 'Ocorreu um erro ao fazer cadastro, tente novamente',
          title: 'Erro no cadastro',
        });
      }
    },
    [addToast, history]
  );

  return (
    <Container>
      <GoBackStyled onClick={history.goBack}>
        <img src={backIcon} alt="back icon" />
      </GoBackStyled>
      <main>
        <figure>
          <img src={logo} alt="logo" />
        </figure>
        <h1>Seja bem vindo.</h1>
      </main>
      <Form ref={formRef} onSubmit={handleSubmit}>
        <section>
          <section>
            <label htmlFor="email">e-mail</label>
            <Input name="email" placeholder="exemplo@gmail.com" />
          </section>

          <section>
            <label htmlFor="confirmEmail">confirme o e-mail</label>
            <Input name="confirmEmail" placeholder="exemplo@gmail.com" />
          </section>

          <section>
            <label htmlFor="password">password</label>
            <Input name="password" type="password" placeholder="******" />
          </section>
        </section>

        <ButtonSection>
          <Button>Sign Up</Button>
        </ButtonSection>
      </Form>
    </Container>
  );
};

export default SignUp;
