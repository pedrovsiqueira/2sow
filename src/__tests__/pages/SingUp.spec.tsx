import React from 'react';
import SignUp from '../../pages/SignUp/SignUp';
import { render, fireEvent, wait } from '@testing-library/react';

const mockedHistoryPush = jest.fn();
const mockedAddToast = jest.fn();

//Mock atribuir valores fictícios para variáveis/funções
//jest.fn é uma função vazia que serve só para saber se ela foi chamada ou não

jest.mock('../../hooks/toast.tsx', () => {
  return {
    useToast: () => ({
      addToast: mockedAddToast,
    }),
  };
});

jest.mock('react-router-dom', () => {
  return {
    useHistory: () => ({
      goBack: jest.fn(),
      push: mockedHistoryPush,
    }),
  };
});

jest.mock('../../services/api.ts', () => {
  return {
    api: {
      post: jest.fn(),
    },
  };
});

describe('SignUp Page', () => {
  beforeEach(() => {
    mockedHistoryPush.mockClear();
  });

  it('renders without crashing', () => {
    const { asFragment } = render(<SignUp />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should be able to sign up', async () => {
    const { getByTestId } = render(<SignUp />);

    const emailField = getByTestId('email');
    const confirmEmailField = getByTestId('confirm-email');
    const passwordField = getByTestId('password');
    const buttonElement = getByTestId('sign-up');

    fireEvent.change(emailField, {
      target: { value: 'pedrosiqueira@gmail.com' },
    });

    fireEvent.change(confirmEmailField, {
      target: { value: 'pedrosiqueira@gmail.com' },
    });

    fireEvent.change(passwordField, {
      target: { value: '123456' },
    });

    fireEvent.click(buttonElement);

    await wait(() => {
      expect(mockedHistoryPush).toHaveBeenCalledWith('/login');
    });
  });

  it('should return an error trying to signUp with invalid credentials', async () => {
    const { getByTestId } = render(<SignUp />);

    const emailField = getByTestId('email');
    const confirmEmailField = getByTestId('confirm-email');
    const passwordField = getByTestId('password');
    const buttonElement = getByTestId('sign-up');

    fireEvent.change(emailField, {
      target: { value: 'not-valid-invalid' },
    });

    fireEvent.change(confirmEmailField, {
      target: { value: 'not-valid-invalid' },
    });

    fireEvent.change(passwordField, {
      target: { value: 'notvalid' },
    });

    fireEvent.click(buttonElement);

    await wait(() => {
      expect(mockedHistoryPush).not.toHaveBeenCalled();
    });
  });
});
