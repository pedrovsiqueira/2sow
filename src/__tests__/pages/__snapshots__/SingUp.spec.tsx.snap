// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp Page renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="sc-AxheI hBbedP"
  >
    <div
      class="sc-Axmtr kGFezI"
    >
      <img
        alt="back icon"
        src="back-icon.svg"
      />
    </div>
    <main>
      <figure>
        <img
          alt="logo"
          src="logo.svg"
        />
      </figure>
      <h1>
        Seja bem vindo.
      </h1>
    </main>
    <form>
      <section>
        <section>
          <label
            class="sc-AxhUy ifxrsp"
            for="email"
          >
            email
          </label>
          <div
            class="sc-AxirZ cmRqOQ"
          >
            <input
              class="sc-AxiKw foiNIc"
              data-testid="email"
              placeholder="exemplo@gmail.com"
            />
          </div>
        </section>
        <section>
          <label
            class="sc-AxhUy ifxrsp"
            for="confirmEmail"
          >
            confirmar email
          </label>
          <div
            class="sc-AxirZ cmRqOQ"
          >
            <input
              class="sc-AxiKw foiNIc"
              data-testid="confirm-email"
              placeholder="exemplo@gmail.com"
            />
          </div>
        </section>
        <section>
          <label
            class="sc-AxhUy ifxrsp"
            for="password"
          >
            senha
          </label>
          <div
            class="sc-AxirZ cmRqOQ"
          >
            <input
              class="sc-AxiKw foiNIc"
              data-testid="password"
              placeholder="******"
              type="password"
            />
          </div>
        </section>
      </section>
      <div
        class="sc-AxmLO bUfhmq"
      >
        <button
          class="sc-AxgMl jrdwSE"
          data-testid="sign-up"
        >
          cadastrar
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
